src/full/Agda/TypeChecking/Monad/Base.hs

```diff
@@ -2677,12 +2728,491 @@ ifTopLevelAndHighlightingLevelIs ::
+
+newtype Het (side :: HetSide) t = Het { unHet :: t }
+  deriving (Foldable, Traversable, Pretty)
+
+deriving instance (Typeable side, Data t) => Data (Het side t)
+deriving instance Show t => Show (Het side t)
+deriving instance Functor (Het side)
+
+deriving instance AllMetas a => AllMetas (Het side a)
+deriving instance Sized a => Sized (Het side a)
+
+instance Decoration (Het s) where
+  traverseF f = fmap coerce . f . coerce
+  distributeF = traverseF id
+
+-- | Converse of `distributeF` for `Het`
+pullHet :: Coercible (f (Het side a)) (f a) => f (Het side a) -> Het side (f a)
+pullHet = coerce
+
+instance Applicative (Het s) where
+  pure = Het
+  Het f <*> Het a = Het (f a)
+
+instance Monad (Het s) where
+  Het a >>= f = f a
+
+instance (HetSideIsType side, Free a) => Free (Het side a) where
+  freeVars' (Het a) = freeVars' a
+
+instance TermLike a => TermLike (Het side a) where
+  foldTerm f = foldTerm f . unHet
+
+instance Apply a => Apply (Het s a) where
+  apply t as = fmap (flip apply as) t
+  applyE t es = fmap (flip applyE es) t
+
+instance Suggest a => Suggest (Het s a) where
+  suggestName = suggestName . twinAt @s
+
+-- * Converting to and from twin types
+
+class AsTwin b where
+  type AsTwin_ b
+  asTwin :: AsTwin_ b -> b
+instance AsTwin b => AsTwin (CompareAs' b) where
+  type AsTwin_ (CompareAs' b) = CompareAs' (AsTwin_ b)
+  asTwin = fmap asTwin
+instance AsTwin (TwinT'' b a) where
+  type AsTwin_ (TwinT'' b a) = a
+  asTwin = SingleT . Het @'Both
+instance AsTwin ContextHet where
+  type AsTwin_ ContextHet = Context
+  asTwin = contextHetFromList . (fmap (fmap (fmap asTwin)))
+instance AsTwin (Het side a) where
+  type AsTwin_ (Het side a) = a
+  asTwin = coerce
+instance AsTwin a => AsTwin (Dom a) where
+  type AsTwin_ (Dom a) = Dom (AsTwin_ a)
+  asTwin = fmap asTwin
+instance AsTwin () where type AsTwin_ () = (); asTwin = id
+instance AsTwin a => AsTwin (Abs a) where
+  type AsTwin_ (Abs a) = Abs (AsTwin_ a)
+  asTwin = fmap asTwin
+instance AsTwin a => AsTwin (Tele a) where
+  type AsTwin_ (Tele a) = Tele (AsTwin_ a)
+  asTwin = fmap asTwin
+instance (AsTwin a, AsTwin b) => AsTwin (a, b) where
+  type AsTwin_ (a, b) = (AsTwin_ a, AsTwin_ b)
+  asTwin (a, b) = (asTwin a, asTwin b)
+instance AsTwin Name where
+  type AsTwin_ Name = Name
+  asTwin = id
+
+class TwinAt (s :: HetSide) a where
+  type TwinAt_ s a
+  twinAt :: a -> TwinAt_ s a
+
+instance HetSideIsType s => TwinAt s (TwinT' a) where
+  type TwinAt_ s (TwinT' a) = a
+  {-# INLINE twinAt #-}
+  twinAt (SingleT a) = unHet @'Both a
+  twinAt TwinT{direction,twinPid,twinLHS,twinRHS} = case (sing :: SingT s) of
+    SLHS    -> unHet @s $ twinLHS
+    SRHS    -> unHet @s $ twinRHS
+    SBoth   | ISet.null twinPid ->
+                selectSmaller direction (unHet @'LHS twinLHS) (unHet @'RHS twinRHS)
+            | otherwise -> __IMPOSSIBLE__
+    SCompat -> __UNIMPLEMENTED__
+
+instance TwinAt s a => TwinAt s (Name, a) where
```

Comment:
Maybe via `default` instances for `Decoration`?

Reply:
Perhaps… But the type family `TwinAt_` complicates things.